### PR TITLE
SLVS-721 Unused private types or members should be removed

### PR DESCRIPTION
**Pull request by tvsonar**

### CODE REVIEW

File: its/expected/Nancy/Nancy-{34576216-0DCA-4B0F-A0DC-9075E75A676F}-S1144.json [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+"issues":  [
+{
+"ruleId":  "S1144",
+"locations":  [
+{
+"analysisTarget":  [
+{
+"uri":  "Nancy\\src\\Nancy\\Conventions\\StaticContentConventionBuilder.cs",
+"region":  {
+"startLine":  245,
```
Comment (tvsonar) [inline its/expected/Nancy/Nancy-{34576216-0DCA-4B0F-A0DC-9075E75A676F}-S1144.json]: Unused property.

File: its/expected/Nancy/Nancy-{34576216-0DCA-4B0F-A0DC-9075E75A676F}-S1144.json [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+},
+{
+"ruleId":  "S1144",
+"locations":  [
+{
+"analysisTarget":  [
+{
+"uri":  "Nancy\\src\\Nancy\\Json\\JsonSerializer.cs",
+"region":  {
+"startLine":  407,
```
Comment (tvsonar) [inline its/expected/Nancy/Nancy-{34576216-0DCA-4B0F-A0DC-9075E75A676F}-S1144.json]: Unused generic method.

File: src/SonarLint.CSharp/Rules/UnusedPrivateMember.cs [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+            context.RegisterCompilationStartAction(
+                csa =>
+                {
+                    var usedSymbols = new HashSet<ISymbol>();
+                    var declaredPrivateSymbols = new HashSet<ISymbol>();
+                    var privateEmptyConstructors = new HashSet<ISymbol>();
+                    var fieldLikeSymbols = new BidirectionalDictionary<ISymbol, SyntaxNode>();
+
+                    RegisterActionForUsedSymbols(csa, usedSymbols);
+                    RegisterActionForConstructorDeclarations(csa, usedSymbols, privateEmptyConstructors);
```
Comment (dbolkensteyn) [inline src/SonarLint.CSharp/Rules/UnusedPrivateMember.cs]: these method names are too technical IMO - I'm more interested what they are supposed to do from a functional point of view than a technical one, to allow me to get the big picture of this rule only be looking at this `RegisterCompilationStartAction`

**Comment by dbolkensteyn** [comment]
looks like every of our new rule is a bit slow, the unused method parameter also takes a few seconds to run

**Comment by dbolkensteyn** [comment]
LGTM

**Pull request merged**